### PR TITLE
WIP: Add Schedule Changes section

### PR DESCRIPTION
**Pull request by e-lo**
## Proposal

This Best Practices proposal addresses the need to understand what needs to happen when a schedule changes unexpectedly and where that should be represented in various parts of GTFS Realtime feeds.  There are potentially multiple ways to address many service situations and this PR should seek to reduce ambiguity so that GTFS Providers can work towards the solution that gives the best information to their customers. 

Addresses part of #5
https://github.com/google/transit/issues/316

## Consensus Building

- [ ] Do we generally agree that this a useful section to have?
- [ ] Do we have alignment on the issues we need to be explicit about?
- [ ] Are we generally aligned recommendations which would benefit the most customers and fit with the policies and procedures of most gtfs-consumers and gtfs-producers?
- [ ] Vote

## Outstanding Issues

To be discussed in PR thread (please propose more if you see them):

## Use Cases

Demonstrate the Best Practice proposal through real-life use cases. Use cases should cover a diversity of transit systems. At least 3 use cases should be provided.

Situations:

 - Temporary construction makes a transit stop unusable.
 - A street fair closes down a block of roadway and the re-routed bus is not able to stop at its usual stop.
 - There's a big snow-storm and some stops are not usable / snow service is being operated

Desired Behavior:

- When a customer plans a trip which is effected by the schedule change, they receive travel options which take into account detours, closures, etc and a **single** alert that this route/stop/trip is effected by detours/closures, etc. 
- When a customer plans a trip which is adjacent to the schedule change (tempo-geographically), they receive the **most relevant alerts** should they need to make small alterations to their plans (i.e. take a later trip, get off one stop later). This should include stops  and trips before and after the planned trip. 
- When a customer plans a trip elsewhere on the route, the irrelevant alert is not displayed. Customers do not need to parse a long list of alerts that may not impact them.
- When a customer requests real-time information for closed stop_id, or is at a stop with a real-time sign configured to display information for that stop_id, they get the relevant alert information. 

cc: @lauramatson who proposed many of these in https://github.com/google/transit/issues/316

### CODE REVIEW

**Comment by e-lo** [comment]
### 2. How do you reference stops/routes etc in an alert which are not in active service for static GTFS?

POV: you update your GTFS Schedule to represent a detour, but want your customers to see an alert when they open their app to their "favorite stop" (as one of many examples). 

**Option A:** Leave stop/route without scheduled service in GTFS Schedule 
- @isabelle-dr would this pass the GTS Schedule validator? 
- @lauramatson where were you receiving pushback on this approach?)

**Option B:** Remove stop/route without scheduled service from GTFS Schedule. Add an alert that doesn't reference GTFS Schedule.  

- @barbeau @isabelle-dr would this pass the RT validator?  there doesn't seem to be anything [in spec itself ](https://gtfs.org/realtime/feed-entities/service-alerts/#entityselector)declaring that it is necessary so much as they should have "GTFS Identifiers" which is poorly defined.
> Entities are selected using their GTFS identifiers

**Option C:** Adopt [GTFS-Service Changes  proposal](https://docs.google.com/document/d/1oPfQ6Xvui0g3xiy1pNiyu5cWFLhvrTVyvYsMVKGyGWM/edit#heading=h.mm6bblup6h8d)
cc GTFS Consumers for what they would accept: @gcamp (TransitApp) @bdferris (G) @barbeau (OTP) @mads14 (Swiftly) and others?

See more on this issue from @lauramatson in https://github.com/google/transit/issues/316 w/ discussion from @sam-hickey-ibigroup @evantrillium @skinkie
> Since it is still a detour (even though it’s in the schedule), we need to provide the Alert information. The challenge we have is that the alert gets updated so that the informed_entity is the route, since the closed stops are no longer referenced in stop_times.txt.
> 
> We tried retaining the stops in stops.txt while removing them from stop_times.txt to keep the alert more targeted, but that was a deal-breaker for a major consumer using our realtime feeds at all. If there was consensus on that approach it would solve part of our issue – it’s hard to parse a long list of route-level alerts – but it would leave some gaps.

**Comment by e-lo** [comment]
1. Do gtfs-producers need to post changes to both Alerts and TripUpdates? Should they? Do they both have an effect on routing?

POV: you can't update your GTFS Schedule to represent a detour, but want customers to know about it in their trip planning and via alerts (as specified in desired behavior above)

| **Situation** | **Alerts** | **TripUpdates** |
| ------------ | ---------- | ----------|
| Routes which are detoured or skipping stops |  **route-level**:  `DETOUR` effect, relevant <br>**stop-level:** `STOP_MOVED` `REDUCED_SERVICE` or `NO-SERVICE` effects |  [`TripUpdates.StopTimeUpdate`](https://gtfs.org/realtime/reference/#message-stoptimeupdate) messages for effected stops. <br>**Missed Stops:**  [`schedule_relationship`](https://gtfs.org/realtime/reference/#enum-schedulerelationship) `SKIPPED` <br> **Added Stops:** [`schedule_relationship`](https://gtfs.org/realtime/reference/#enum-schedulerelationship) `UNSCHEDULED` |
| Skipped trips |  **route-level**:  `REDUCED_SERVICE`<br> **trip-level**:  `No service` | [`TripUpdates.StopTimeUpdate`](https://gtfs.org/realtime/reference/#message-stoptimeupdate) message for effected trips.<br> [`schedule_relationship`](https://gtfs.org/realtime/reference/#enum-schedulerelationship)  `SKIPPED` |
| Station entrances/exits closed | **stop-level:** `NO-SERVICE` effect. | None? |

Question for gtfs-consumers:  @gcamp (TransitApp) @bdferris (G) @barbeau (OTP) @mads14 (Swiftly) and others? 

Is one or the other necessary and sufficient to achieve the above desired behavior? Ideally we can have a single method of accomplishing an objective.

**Comment by gcamp** [comment]
1. Both options are bad in my opinion. A. would be filtered in the preprocessing most likely and B. if there's no target the alert would just not be shown anywhere. I would propose something that describe the detour explicitly like service changes proposes, I don't think there's a good way to do it now. 

2. We modify routing/behavior of the app with `NO_SERVICE` alerts, but I know spec-wise it has never been entirely agreed on. No other alerts change behaviour as the effect is specific enough to be acted upon.

Changes should always be in TripUpdates and Alerts should be present if the agency wants a user-visible message. In the "Situations" above, all should have both alerts and trip updates. I understand the goal of having a single source of truth but I don't think it's possible without redesigning completely alerts and TripUpdates.

**Comment by e-lo** [comment]
@gcamp 
> I would propose something that describe the detour explicitly like service changes proposes, I don't think there's a good way to do it now.

I added adoption of this as **Option C**, but it is troubling that there is no way to represent this currently as there is likely a significant lift to adopting this extension (please correct me if I'm wrong here)

**Comment by martaswiftly** [comment]
Hello everyone,

I’m responding on behalf of Swiftly after syncing with Maddie.

### Question 1:  Do gtfs-producers need to post changes to both Alerts and TripUpdates? Should they? Do they both have an effect on routing?

At Swiftly, our current school of thought is that riders should have the right amount of information at their fingertips to make an informed decision without feeling overwhelmed. Given that, we're always willing to learn more and update our practices based on input from the community.

Depending on the different situations, the information a rider would need might differ. 
While our approach is to reflect changes in TripUpdates (to inform riders and keep a historical record of the changes), we find that deciding when Service Alerts should be published is a little bit more nuanced. A few examples are described here:

1. There is a vehicle that broke down and a trip will be delayed 7 minutes. An agency might want to skip the alert for the delayed trip due to the small delay (that is already informed via TripUpdates).
2. A trip is canceled and the next one will arrive in 1 hour: in that situation, a service alert might seem appropriate to inform riders of the reduced service. 
3. A trip is canceled and the next one will arrive in 6 minutes: a Service Alert might be perceived as noisy by riders. The TripUpdates entity might be enough in this case.
4. There is a snowstorm and more than 50% of the routes will be removed, and severe delays are expected. In this case, an agency might want to share a general alert with riders to let them know of the unexpected weather conditions and the reduced service including the list of routes that are removed. 

Agencies have the full context of the situation and know the rider severity of the service changes as they occur, so that’s why we believe agencies are best positioned to decide how many alerts (if any) are necessary.

While providing the best rider experience is the common goal, we recommend not overwhelming subscribed riders with duplicative alerts due to the same event. Riders just need the minimum amount of notifications to be well informed. 

When creating alerts for passengers, these can impact a stop across all routes or a stop within a specific route or set of routes. See [examples](https://gtfs.org/realtime/feed-examples/service-alerts/). Swiftly generates an alert where both `routeid` and `stopid` are referenced in ServiceAlerts.informedEntities. In this case, both entities are referenced to the same effect. 

Please, find our [suggested approach](https://docs.google.com/document/d/1Xt4p7vMY8hLbDcsYr9Uks_WLxD4NaZi0SMTa34bSBho/edit?usp=sharing) for the stated detour situations in more detail.


### Question 2 How do you reference stops/routes etc in an alert which are not in active service for static GTFS?

To inform riders about a detour that was created in the static GTFS, we recommend referencing the alert to the route-level. That way, riders will get notified when interacting with the route of that planned detour. No need to reference the `stopid` as it's no longer active.

**Comment by isabelle-dr** [comment]
@e-lo to your specific questions about the Canonical Validators

> Option A: Leave stop/route without scheduled service in GTFS Schedule
> @isabelle-dr would this pass the GTS Schedule validator?

You'd get [stop_without_stop_time](https://github.com/MobilityData/gtfs-validator/blob/master/RULES.md#stop_without_stop_time) (warning), but there is no explicit mention of that in the sec. It's more an FYI than a violation. The data would still be valid.

> Option B: Remove stop/route without scheduled service from GTFS Schedule. Add an alert that doesn't reference GTFS Schedule.
> @barbeau @isabelle-dr would this pass the RT validator? there doesn't seem to be anything [in spec itself ](https://gtfs.org/realtime/feed-entities/service-alerts/#entityselector)declaring that it is necessary so much as they should have "GTFS Identifiers" which is poorly defined.

I have less expertise there, but after having a look at the list of rules, it seems it would pass, yes.

**Comment by Sergiodero** [comment]
As GTFS Best Practices (BP) are currently in the process of being merged to the specification, MobilityData is migrating outstanding issues and PRs from this repository to google/transit. Thus, this PR will be closed and further discussion regarding this BP should take place in google/transit. Please refer to [Issue #421](https://github.com/google/transit/issues/421) for a more detailed explanation of the migration process and the proposed next steps. 

With this, we’re hoping to bring more visibility to outstanding BP issues and to restart the discussion around them, so that any improvement that the community finds valuable could be carried forward.
